src/mjolnir/adminbuilder.cc

```diff
@@ -12,206 +12,175 @@
-#if GEOS_VERSION_MAJOR == 3 && GEOS_VERSION_MINOR >= 9
-  for (unsigned i = 0; i < merged.size(); ++i) {
-    std::unique_ptr<LineString> pline(merged[i].release());
-#else
-  for (unsigned i = 0; i < merged->size(); ++i) {
-    std::unique_ptr<LineString> pline((*merged)[i]);
-#endif
-    if (pline->getNumPoints() > 3 && pline->isClosed()) {
-#if GEOS_VERSION_MAJOR == 3 && GEOS_VERSION_MINOR >= 8
-      polys[totalpolys].polygon =
-          gf->createPolygon(gf->createLinearRing(pline->getCoordinates())).release();
-      polys[totalpolys].ring = gf->createLinearRing(pline->getCoordinates()).release();
-#else
-      polys[totalpolys].polygon = gf->createPolygon(gf->createLinearRing(pline->getCoordinates()), 0);
-      polys[totalpolys].ring = gf->createLinearRing(pline->getCoordinates());
-#endif
-      polys[totalpolys].area = polys[totalpolys].polygon->getArea();
-      polys[totalpolys].iscontained = 0;
-      polys[totalpolys].containedbyid = 0;
-      if (polys[totalpolys].area > 0.0) {
-        totalpolys++;
-      } else {
-        delete (polys[totalpolys].polygon);
-        delete (polys[totalpolys].ring);
-      }
-    }
+  std::vector<polygon_data> polys;
+  for (const auto& ring : rings) {
+    polygon_data pd{};
+    boost::geometry::assign_points(pd.polygon, ring);
+    boost::geometry::correct(pd.polygon);
+    if (!boost::geometry::is_valid(pd.polygon))
+      continue;
+    pd.area = boost::geometry::area(pd.polygon);
+    polys.emplace_back(std::move(pd));
   }
 
-  if (totalpolys) {
-    qsort(polys, totalpolys, sizeof(polygondata), polygondata_comparearea);
+  // if we have anything that wasnt degenerate we can convert it to wkt
+  if (polys.empty())
+    return "";
 
-    for (unsigned i = 0; i < totalpolys; ++i) {
-      if (polys[i].iscontained != 0) {
-        continue;
-      }
+  // sort descending by area
+  std::sort(polys.begin(), polys.end());
 
-      for (unsigned j = i + 1; j < totalpolys; ++j) {
-        // Does polygon[i] contain the smaller polygon[j]?
-        if (polys[j].containedbyid == 0 && polys[i].polygon->contains(polys[j].polygon)) {
-          // are we in a [i] contains [k] contains [j] situation
-          // which would actually make j top level
-          bool istoplevelafterall = false;
-          for (unsigned k = i + 1; k < j; ++k) {
-            if (polys[k].iscontained && polys[k].containedbyid == i &&
-                polys[k].polygon->contains(polys[j].polygon)) {
-              istoplevelafterall = true;
-              break;
-            }
-          }
-          if (istoplevelafterall) {
-            polys[j].iscontained = 1;
-            polys[j].containedbyid = i;
-          }
-        }
-      }
+  // we need to figure out which rings are inners of others
+  // our strategy is to only ever go one level deep with inners, that is inners of inners
+  // will just become outers that happen to be inside the hole of another outer
+  for (size_t i = 0; i < polys.size(); ++i) {
```

Comment:
this algorithm in master looks completely broken to me. it looks to me like inners shouldnt work at all in that version because of that "toplevelafterall" variable being set to the opposite of what it should be. there is also some strange stuff with checking the "contained_by_id" being 0 where i think it really meant to be checking "is_contained". anyway at least the unit tests proved this is working for fake data. im testing with the planet now, we'll see how it goes

Reply:
haha yeah I had the exact same thought back when I tried. no idea how enclaves test was even working

Reply:
there was no enclaves unit tests i just added one. that said i checked the output of and old build and they were working in that baarle hertog example looked great

Reply:
ah I thought it had one.. thanks for adding one